File: lib/java/opentoken/src/main/java/com/truveta/opentoken/tokens/definitions/T5Token.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokens.definitions;

import java.util.ArrayList;
import com.truveta.opentoken.attributes.AttributeExpression;
import com.truveta.opentoken.attributes.person.FirstNameAttribute;
import com.truveta.opentoken.attributes.person.LastNameAttribute;
import com.truveta.opentoken.attributes.person.SexAttribute;
import com.truveta.opentoken.tokens.Token;

/**
 * Represents the token definition for token T5.
 * 
 * <p>
 * It is a collection of attribute expressions
 * that are concatenated together to get the token signature.
 * The token signature is as follows:
 * U(last-name)|U(first-name-3)|U(gender)
 * </p>
 * 
 * @see com.truveta.opentoken.tokens.Token Token
 */
public class T5Token implements Token {
    private static final String ID = "T5";

    private final ArrayList<AttributeExpression> definition = new ArrayList<>();

    public T5Token() {
        definition.add(new AttributeExpression(LastNameAttribute.class, "T|U"));
        definition.add(new AttributeExpression(FirstNameAttribute.class, "T|S(0,3)|U"));
        definition.add(new AttributeExpression(SexAttribute.class, "T|U"));
    }

    @Override
    public String getIdentifier() {
        return ID;
    }

    @Override
    public ArrayList<AttributeExpression> getDefinition() {
        return definition;
    }
}

File: lib/python/opentoken/src/main/opentoken/tokens/definitions/t5_token.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List
from opentoken.attributes.attribute_expression import AttributeExpression
from opentoken.attributes.person.first_name_attribute import FirstNameAttribute
from opentoken.attributes.person.last_name_attribute import LastNameAttribute
from opentoken.attributes.person.sex_attribute import SexAttribute
from opentoken.tokens.token import Token


class T5Token(Token):
    """
    Represents the token definition for token T5.

    It is a collection of attribute expressions that are concatenated together
    to get the token signature. The token signature is as follows:
    U(last-name)|U(first-name-3)|U(gender)
    """

    ID = "T5"

    def __init__(self):
        """Initialize the T5 token definition."""
        self._definition = [
            AttributeExpression(LastNameAttribute, "T|U"),
            AttributeExpression(FirstNameAttribute, "T|S(0,3)|U"),
            AttributeExpression(SexAttribute, "T|U")
        ]

    def get_identifier(self) -> str:
        """Get the unique identifier for the token."""
        return self.ID

    def get_definition(self) -> List[AttributeExpression]:
        """Get the list of attribute expressions that define the token."""
        return self._definition
